src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/DefaultLSPProgressListener.cs

```diff
@@ -0,0 +1,192 @@
+                // We can deem this ProgressRequest complete
+                ProgressCompleted(token);
+                return;
+            }
+
+            CompleteAfterDelay(token, request);
+        }
+
+        private void CompleteAfterDelay(string token, ProgressRequest request)
+        {
+            CancellationTokenSource linkedCTS;
+
+            lock (request.RequestLock)
+            {
+                var existingTimeoutCTS = request.TimeoutCancellationTokenSource;
+                if (existingTimeoutCTS != null &&
+                    existingTimeoutCTS.Token.CanBeCanceled &&
+                    !existingTimeoutCTS.Token.IsCancellationRequested)
+                {
+                    existingTimeoutCTS.Cancel();
+                }
+
+                request.TimeoutCancellationTokenSource = new CancellationTokenSource();
+                linkedCTS = CancellationTokenSource.CreateLinkedTokenSource(
+                    request.TimeoutCancellationTokenSource.Token,
+                    request.HandlerCancellationToken);
+            }
+
+            _ = CompleteAfterDelayAsync(token, request.TimeoutAfterLastNotify, linkedCTS.Token); // Fire and forget
+        }
+
+        private async Task CompleteAfterDelayAsync(string token, TimeSpan delay, CancellationToken cancellationToken)
+        {
+            try
+            {
+                await Task.Delay(delay, cancellationToken).ConfigureAwait(false);
+            }
+            catch (TaskCanceledException)
+            {
+                // Task cancelled, new progress notification received.
+                // Don't allow handler to return
+                return;
+            }
+
+            ProgressCompleted(token);
+        }
+
+        public override bool TryListenForProgress(
+            string token,
+            Func<JToken, Task> onProgressNotifyAsync,
+            TimeSpan timeoutAfterLastNotify,
+            CancellationToken handlerCancellationToken,
+            out Task onCompleted)
+        {
+            var onCompletedSource = new TaskCompletionSource<bool>();
+            var request = new ProgressRequest(
+                onProgressNotifyAsync,
+                timeoutAfterLastNotify,
+                handlerCancellationToken,
+                onCompletedSource);
+
+            if (!_activeRequests.TryAdd(token, request))
+            {
+                onCompleted = null;
+                return false;
+            }
+
+            CompleteAfterDelay(token, request);
+            onCompleted = onCompletedSource.Task;
+            return true;
+        }
+
+        private void ProgressCompleted(string token)
+        {
+            if (_activeRequests.TryRemove(token, out var request))
+            {
+                // We're setting the result of a Task<bool>
+                // however we only return a Task to the
+                // handler/subscriber, so the bool is ignored
+                request.OnCompleted.SetResult(false);
```

Comment:
CTS' can leak so when you remove the active request you also need to clean up all of its cancellation token state.